src/backend/distributed/utils/acquire_lock.c

```diff
@@ -0,0 +1,311 @@
+	}
+
+
+	MemoryContextCallback *cb = palloc0(sizeof(MemoryContextCallback));
+	cb->func = EnsureStopLockAcquireHelper;
+	cb->arg = handle;
+
+	MemoryContextRegisterResetCallback(CurrentMemoryContext, cb);
+
+	return handle;
+}
+
+
+/*
+ * EnsureStopLockAcquireHelper is designed to be called as a MemoryContextCallback. It
+ * takes a handle to the background worker and Terminates it. It is safe to be called on a
+ * handle that has already been terminated due to the guard around the generation number
+ * implemented in the handle by postgres.
+ */
+static void
+EnsureStopLockAcquireHelper(void *arg)
+{
+	BackgroundWorkerHandle *handle = (BackgroundWorkerHandle *) arg;
+	TerminateBackgroundWorker(handle);
+}
+
+
+/*
+ * Signal handler for SIGTERM
+ *		Set a flag to let the main loop to terminate, and set our latch to wake
+ *		it up.
+ */
+static void
+lock_acquire_helper_sigterm(SIGNAL_ARGS)
+{
+	int save_errno = errno;
+
+	got_sigterm = true;
+	SetLatch(MyLatch);
+
+	errno = save_errno;
+}
+
+
+/*
+ * ShouldAcquireLock tests if our backend should still proceed with acquiring the lock,
+ * and thus keep terminating conflicting backends. This function returns true until a
+ * SIGTERM, background worker termination signal, has been received.
+ *
+ * The function blocks for at most sleepms when called. During operation without being
+ * terminated this is the time between invocations to the backend termination logic.
+ */
+static bool
+ShouldAcquireLock(long sleepms)
+{
+	int rc;
+
+	rc = WaitLatch(MyLatch, WL_LATCH_SET | WL_TIMEOUT | WL_POSTMASTER_DEATH,
+				   sleepms * 1L, PG_WAIT_EXTENSION);
+	ResetLatch(MyLatch);
+
+	/* emergency bailout if postmaster has died */
+	if (rc & WL_POSTMASTER_DEATH)
+	{
+		proc_exit(1);
+	}
+
+	CHECK_FOR_INTERRUPTS();
+
+	return !got_sigterm;
+}
+
+
+/*
+ * LockAcquireHelperMain runs in a dynamic background worker to help master_update_node to
+ * acquire its locks.
+ */
+void
+LockAcquireHelperMain(Datum main_arg)
+{
+	int backendPid = DatumGetInt32(main_arg);
+	StringInfoData sql;
+	LockAcquireHelperArgs *args = (LockAcquireHelperArgs *) MyBgworkerEntry->bgw_extra;
+	long timeout = 0;
+	const TimestampTz connectionStart = GetCurrentTimestamp();
+	const TimestampTz deadline = TimestampTzPlusMilliseconds(connectionStart,
+															 args->lock_cooldown);
+
+	pqsignal(SIGTERM, lock_acquire_helper_sigterm);
+
+	BackgroundWorkerUnblockSignals();
+
+	elog(LOG, "lock acquiring backend started for backend %d (cooldown %dms)", backendPid,
+		 args->lock_cooldown);
+
+	/*
+	 * this loop waits till the deadline is reached (eg. lock_cooldown has passed) OR we
+	 * no longer need to acquire the lock due to the termination of this backend.
+	 * Only after the timeout the code will continue with the section that will acquire
+	 * the lock.
+	 */
+	do {
+		timeout = DeadlineTimestampTzToTimeout(deadline);
+	} while (timeout > 0 && ShouldAcquireLock(timeout));
+
```

Comment:
Maybe add an early exit in case of got_sigterm.

Reply:
`ShouldAcquireLock(...)` escapes on `got_sigterm` albeit not early. We could change it in such a way that before entering the wait we can already check for `got_sigterm`. Or is your point to have an explicit `&& !got_sigterm` in the condition?

Reply:
Either way seems fine, currently it waits 100ms for no particular reason.

Reply:
added an early escape in `ShouldAcquireLock` as it is also used in the loop below.